pandas/indexes/base.py

```diff
@@ -2886,6 +2888,9 @@ def join(self, other, how='left', level=None, return_indexers=False):
         elif how == 'outer':
             join_index = self.union(other)
 
+        if sort:
```

Comment:
I admit there is an ambiguity here. 

By default (`sort=False`) the result index values are not sorted, except for the outer join. This is because Index.union sorts the result index values. 

If we set `sort=True`, the result index values are sorted for all types of join (although this is redundant for the outer join, as it is already sorted).